src/storage/nats_kv.rs

```diff
@@ -0,0 +1,248 @@
+            bucket,
+            history: 10,
+            ..Default::default()
+        }).await
+            .map_err(|e| StorageError::Engine(format!("failed to create jetstream kv: {e}")))?;
+
+        Ok(kv)
+    }
+}
+
+#[async_trait]
+impl EngineMetadata for NatsKvStorageEngine {
+    async fn name() -> String {
+        String::from(STORAGE_ENGINE_NAME)
+    }
+
+    async fn version() -> Version {
+        STORAGE_ENGINE_VERSION
+    }
+}
+
+#[async_trait]
+impl Store for NatsKvStorageEngine {
+    type State = LatticeState;
+    type StateId = String;
+    type StateUpdate = LatticeStateBuilder;
+
+    async fn get(&self, id: String) -> Result<WithStateMetadata<LatticeState>, StorageError> {
+        Ok(self.read(id).await?)
+    }
+
+    async fn store(&mut self, state: LatticeState, store_opts: StoreOptions) -> Result<WithStateMetadata<String>, StorageError> {
+        Ok(self.write(&state, None, store_opts).await?)
+    }
+
+    // Delete an existing piece of state
+    async fn delete(&self, id: String) -> Result<(), StorageError> {
+        // Retrieve the kv store for a given lattice
+        let bucket = self.build_lattice_bucket_name(id);
+        let kv = self.create_kv_store(bucket.clone()).await?;
+
+        let _ = kv.delete(bucket)
+            .await
+            .map_err(|err| StorageError::Engine(format!("failed to delete lattice state: {err}")))?;
+
+        Ok(())
+    }
+}
+
+impl LatticeStorage for NatsKvStorageEngine {}
+
+/////////////////////////////
+// Storage Implementations //
+/////////////////////////////
+
+//////////////////////////
+// Store - LatticeState //
+//////////////////////////
+
+impl From<SerdeJsonError> for StorageError {
+    fn from(e: SerdeJsonError) -> Self {
+        StorageError::Unknown(format!("de/serialization error occurred: {e}"))
+    }
+}
+
+impl From<Box<dyn StdError + std::marker::Send + Sync>> for StorageError {
+    fn from(e: Box<dyn StdError + std::marker::Send + Sync>) -> Self {
+        StorageError::Unknown(format!("de/serialization error occurred: {e}"))
+    }
+}
+
+impl From<LatticeStateBuilderError> for StorageError {
+    fn from(e: LatticeStateBuilderError) -> Self {
+        StorageError::Unknown(format!("failed to build lattice: {e}"))
+    }
+}
+
+const LATTICE_BUCKET_STATE_KEY: &str = "state";
+
+/// Implementation of lattice state storage for NATS KV
+///
+/// For every lattice this implementation stores the all lattice state under a single key
+/// determined by LATTICE_BUCKET_STATE_KEY.
+/// For example, a lattice bucket "lattice_1", you can find serialized lattice state at "lattice_1.state".
+#[async_trait]
+impl StoreRead<LatticeState,String> for NatsKvStorageEngine {
+    async fn read(&self, id: String) -> Result<WithStateMetadata<LatticeState>, StorageError> {
+        let bucket = self.build_lattice_bucket_name(id);
+        let kv = self.create_kv_store(bucket.clone()).await?;
+
+        let lattice;
```

Comment:
does Clippy actually let you get away with this? I remember doing this once and Clippy got super mad at me

Reply:
Yup! :) I added the `make lint` just so I could clippy. `fmt` has a bunch of complaints but clippy is happy